rasa/core/actions/forms.py

```diff
@@ -597,49 +588,60 @@ def _should_request_slot(tracker: "DialogueStateTracker", slot_name: Text) -> bo
-        - submit the form if all required slots are set
-        - deactivate the form
-        """
+        # collect values of required slots filled before activation
+        prefilled_slots = {}
 
-        self._domain = domain
+        for slot_name in self.required_slots(domain):
+            if not self._should_request_slot(tracker, slot_name):
+                prefilled_slots[slot_name] = tracker.get_slot(slot_name)
 
-        # activate the form (we don't return these events in case form immediately
-        # finishes)
-        events = await self._activate_if_required(tracker, domain, output_channel, nlg)
-        # validate user input
-        events += await self._validate_if_required(tracker, domain, output_channel, nlg)
-        # check that the form wasn't deactivated in validation
-        if Form(None) not in events:
-            temp_tracker = self._temporary_tracker(tracker, events, domain)
+        if not prefilled_slots:
+            logger.debug("No pre-filled required slots to validate.")
+            return []
 
-            next_slot_events = await self.request_next_slot(
-                temp_tracker, domain, output_channel, nlg
-            )
+        logger.debug(f"Validating pre-filled required slots: {prefilled_slots}")
+        return await self.validate_slots(
+            prefilled_slots, tracker, domain, output_channel, nlg
+        )
 
-            if next_slot_events is not None:
-                # request next slot
-                events += next_slot_events
-            else:
-                # there is nothing more to request, so we can submit
-                self._log_form_slots(temp_tracker)
-                # deactivate the form after submission
-                events.extend(self.deactivate())  # type: ignore
+    async def do(
+        self,
+        output_channel: "OutputChannel",
+        nlg: "NaturalLanguageGenerator",
+        tracker: "DialogueStateTracker",
+        domain: "Domain",
+        events_so_far: List[Event],
+    ) -> List[Event]:
+        events = await self._validate_if_required(tracker, domain, output_channel, nlg)
+
+        temp_tracker = self._temporary_tracker(tracker, events_so_far + events, domain)
+        events += await self.request_next_slot(
+            temp_tracker, domain, output_channel, nlg
+        )
 
         return events
 
-    def __str__(self) -> Text:
-        return f"FormAction('{self.name()}')"
+    async def is_done(
+        self,
+        output_channel: "OutputChannel",
+        nlg: "NaturalLanguageGenerator",
+        tracker: "DialogueStateTracker",
+        domain: "Domain",
+        events_so_far: List[Event],
+    ) -> bool:
+        # TODO: This will result in a changed order. Is that okay?
+        return SlotSet(REQUESTED_SLOT, None) in events_so_far
+
+    async def deactivate(self, *args: Any, **kwargs: Any) -> List[Event]:
+        logger.debug(f"Deactivating the form '{self.name()}'")
+        return []
```

Comment:
should this return `Form(None)`?

Reply:
aaah, this is done by loop action

Reply:
Yep, since it's the common deactivation behavior for all `loop`s